sys/riscv/trap.c

```diff
@@ -0,0 +1,301 @@
+
+static __noreturn void kernel_oops(ctx_t *ctx) {
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+
+  klog("%s at %p!", exceptions[code], epc);
+
+  switch (code) {
+    case SCAUSE_INST_MISALIGNED:
+    case SCAUSE_INST_ACCESS_FAULT:
+    case SCAUSE_LOAD_MISALIGNED:
+    case SCAUSE_LOAD_ACCESS_FAULT:
+    case SCAUSE_STORE_MISALIGNED:
+    case SCAUSE_STORE_ACCESS_FAULT:
+    case SCAUSE_INST_PAGE_FAULT:
+    case SCAUSE_LOAD_PAGE_FAULT:
+    case SCAUSE_STORE_PAGE_FAULT:
+      klog("Caused by reference to %lx!", _REG(ctx, TVAL));
+      break;
+
+    case SCAUSE_ILLEGAL_INSTRUCTION: {
+      uint32_t badinstr = *(uint32_t *)epc;
+      klog("Illegal instruction %08x in kernel mode!", badinstr);
+      break;
+    }
+
+    case SCAUSE_BREAKPOINT:
+      klog("No debbuger in kernel!");
+      break;
+  }
+  klog("HINT: Type 'info line *%p' into gdb to find faulty code line", epc);
+
+  panic("KERNEL PANIC!!!");
+}
+
+static void page_fault_handler(ctx_t *ctx) {
+  thread_t *td = thread_self();
+
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+  vaddr_t vaddr = _REG(ctx, TVAL);
+
+  klog("%s at %p, caused by reference to %lx!", exceptions[code], epc, vaddr);
+
+  pmap_t *pmap = pmap_lookup(vaddr);
+  if (!pmap) {
+    klog("No physical map defined for %lx address!", vaddr);
+    goto fault;
+  }
+
+  vm_prot_t access;
+  if (code == SCAUSE_INST_PAGE_FAULT)
+    access = VM_PROT_EXEC;
+  else if (code == SCAUSE_LOAD_PAGE_FAULT)
+    access = VM_PROT_READ;
+  else
+    access = VM_PROT_WRITE;
+
+  int error = pmap_emulate_bits(pmap, vaddr, access);
+  if (!error)
+    return;
+
+  if (error == EACCES || error == EINVAL)
+    goto fault;
+
+  vm_map_t *vmap = vm_map_lookup(vaddr);
+  if (!vmap) {
+    klog("No virtual address space defined for %lx!", vaddr);
+    goto fault;
+  }
+
+  if (!vm_page_fault(vmap, vaddr, access))
+    return;
+
+fault:
+  if (td->td_onfault) {
+    /* Handle copyin/copyout faults. */
+    _REG(ctx, PC) = td->td_onfault;
+  } else if (user_mode_p(ctx)) {
+    /* Send a segmentation fault signal to the user program. */
+    sig_trap(ctx, SIGSEGV);
+  } else {
+    /* Panic when kernel-mode thread uses wrong pointer. */
+    kernel_oops(ctx);
+  }
+}
+
+/*
+ * RISC-V syscall ABI:
+ *  - a7: code
+ *  - a0-5: args
+ *
+ * NOTE: the following code assumes all arguments to syscalls are passed
+ * via registers.
+ */
+static_assert(SYS_MAXSYSARGS <= FUNC_MAXREGARGS - 1,
+              "Syscall args don't fit in registers!");
+
+static void syscall_handler(mcontext_t *uctx, syscall_result_t *result) {
+  /* TODO: eventually we should have a platform-independent syscall handler. */
+  register_t args[SYS_MAXSYSARGS];
+  register_t code = _REG(uctx, A7);
+
+  memcpy(args, &_REG(uctx, A0), sizeof(args));
+
+  if (code > SYS_MAXSYSCALL) {
+    args[0] = code;
+    code = 0;
+  }
+
+  sysent_t *se = &sysent[code];
+  size_t nargs = se->nargs;
+
+  assert(nargs <= SYS_MAXSYSCALL);
```

Comment:
I have mixed fillings about this assertion. It's ok to do that here but I think that it would be better to have separate place where we can validate number of arguments for each system call. Why? because here we only validate used syscalls so it's possible that we don't have full code coverage.

Reply:
First of all, there should be `SYS_MAXSYSARGS` instead of `SYS_MAXSYSCALL` :). I agree with you. I will consult it with @cahirwpz to identify where this check should be performed (for all syscall arg structs).

Reply:
While I understand your point I think we should firstly seek to move `syscall_handler` to `sys/gen` directory and make it common for all architectures. Then we'll solve all problems in one place.